src/chainstate/coordinator/mod.rs

```diff
@@ -0,0 +1,486 @@
+    Burnchain, BurnchainBlockHeader,
+    db::{
+        BurnchainDB, BurnchainBlockData
+    }
+};
+use chainstate::burn::{BlockHeaderHash, BlockSnapshot};
+use chainstate::burn::db::sortdb::{SortitionDB, PoxId, SortitionId};
+use chainstate::stacks::{
+    StacksBlock, StacksBlockId, TransactionPayload,
+    Error as ChainstateError, events::StacksTransactionReceipt,
+};
+use chainstate::stacks::db::{
+    StacksHeaderInfo, StacksChainState, ClarityTx
+};
+use monitoring::{
+    increment_stx_blocks_processed_counter,
+};
+use vm::{
+    costs::ExecutionCost,
+    types::PrincipalData
+};
+use util::db::{
+    Error as DBError
+};
+
+#[derive(Debug, PartialEq)]
+pub struct RewardCycleInfo {
+    /// what was the elected PoX anchor, if any?
+    pub anchor_block: Option<BlockHeaderHash>,
+    /// was the elected PoX anchor known?
+    pub anchor_block_known: bool
+}
+
+pub trait BlockEventDispatcher {
+    fn announce_block(&self, block: StacksBlock, metadata: StacksHeaderInfo,
+                      receipts: Vec<StacksTransactionReceipt>, parent: &StacksBlockId);
+}
+
+pub struct ChainsCoordinator<'a, T: BlockEventDispatcher> {
+    canonical_sortition_tip: Option<SortitionId>,
+    canonical_chain_tip: Option<StacksBlockId>,
+    canonical_pox_id: Option<PoxId>,
+    burnchain_blocks_db: BurnchainDB,
+    chain_state_db: StacksChainState,
+    sortition_db: SortitionDB,
+    burnchain: Burnchain,
+    dispatcher: Option<&'a T>
+}
+
+#[derive(Debug)]
+pub enum Error {
+    BurnchainBlockAlreadyProcessed,
+    BurnchainError(BurnchainError),
+    ChainstateError(ChainstateError),
+    NonContiguousBurnchainBlock(BurnchainError),
+    NoSortitions,
+    FailedToProcessSortition(BurnchainError),
+    DBError(DBError),
+}
+
+impl From<BurnchainError> for Error {
+    fn from(o: BurnchainError) -> Error {
+        Error::BurnchainError(o)
+    }
+}
+
+impl From<ChainstateError> for Error {
+    fn from(o: ChainstateError) -> Error {
+        Error::ChainstateError(o)
+    }
+}
+
+impl From<DBError> for Error {
+    fn from(o: DBError) -> Error {
+        Error::DBError(o)
+    }
+}
+
+struct CoordinatorChannels {
+    new_stacks_block_channel: Sender<()>,
+    new_burn_block_channel: Sender<()>,
+}
+
+struct CoordinatorReceivers {
+    event_stacks_block: Receiver<()>,
+    event_burn_block: Receiver<()>,
+}
+
+// Singletons for ChainsCoordinator communication
+//
+//  these channels allow any thread to notify the ChainsCoordinator
+//   instance that a new staging block is ready or a new bitcoin
+//   block has arrived
+//
+//  using a singleton for this pretty dramatically simplifies state
+//   management in the stacks-node, bitcoin indexer, and relayer, because they
+//   don't need to pass around instances of the channels. however,
+//   this _does_ step on the cargo test framework in silly ways, so any
+//   tests which instantiate a coordinator need to call
+//   CoordinatorCommunication::stop_chains_coordinator()
```

Comment:
Good point -- will refactor to make this a little more of a conventional implementation.

Reply:
Updated in 71116e1 -- it still provides the static access functions for now to avoid needing to clone the singleton (the `shared()` method can't return a reference, because RwLock.read() only exposes a temporary reference.